p4-16/psa/PSA.mdk

```diff
@@ -469,12 +510,164 @@ have an instance for each entry in the table.
+        // Packet and byte counts share storage in the same counter
+        // state.  Should we have a separate constructor with an
+        // additional argument indicating how many of the bits to use
+        // for the byte counter?
+        W shift_amount = TBD;
+        this.shifted_packet_count = ((W) 1) << shift_amount;
+        this.packet_count_mask = (~((W) 0)) << shift_amount;
+        this.byte_count_mask = ~this.packet_count_mask;
+    }
+}
+
+W next_counter_value(W cur_value, CounterType_t type) {
+    if (type == CounterType_t.packets) {
+        return (cur_value + 1);
+    }
+    // Exactly which packet bytes are included in packet_len is
+    // implementation-specific.
+    PacketLength_t packet_len = <packet length in bytes>;
+    if (type == CounterType_t.bytes) {
+        return (cur_value + packet_len);
+    }
+    // type must be CounterType_t.packets_and_bytes
+    // In type W, the least significant bits contain the byte
+    // count, and most significant bits contain the packet count.
+    // This is merely one example storage format.  Implementations
+    // are free to store packets_and_byte state in other ways, as
+    // long as the control plane API returns the correct separate
+    // packet and byte count values.
+    W next_packet_count = ((cur_value + this.shifted_packet_count) &
+                           this.packet_count_mask);
+    W next_byte_count = (cur_value + packet_len) & this.byte_count_mask;
+    return (next_packet_count | next_byte_count);
+}
+
+void count(in S index) {
+    if (index < this.num_counters) {
+        this.counter_vals[index] = next_counter_value(this.counter_vals[index],
+                                                      this.type);
+    } else {
+        // No counter_vals updated if index is out of range.
+        // See below for optional debug information to record.
+    }
+}
+```
+
+Optional debugging information that may be kept if an `index` value is
+out of range includes:
+
+- Number of times this occurs.
+- A FIFO of the first N out-of-range index valus that occur, where N
+  is implementation-defined (e.g. it might only be 1).  Extra
+  information to identify which `count()` method call in the P4
+  program had the out-of-range `index` value is also recommended.
+
+
 ### Direct Counter
 
 ```
 [INCLUDE=psa.p4:DirectCounter_extern]
 ```
 
+A `DirectCounter` instance must appear in the list of values of the
+`psa_direct_counters` table attribute for exactly one table.  We call
+this table the `DirectCounter` instance's "owner".  It is illegal to
+call the `count` method for a `DirectCounter` instance anywhere except
+inside an action of its owner table.
+
+The counter value updated by an invocation of `count` is always the
+one associated with the table entry that matched.
+
+TBD: How to describe which counter value is updated for tables with
+action profiles and direct counters?  Or should this combination even
+be allowed?
```

Comment:
Hi Andy, 

Below is a sample v1 model program that uses action selector and direct counter. Direct counters and meters are resources tied to the table entry, and can only be addressed by the match key for that table. They are not tied in any way to action profile members or groups. 

For the foo_table (below), the table entries need to be created after an action profile member/group has been created. P4Runtime provides a TableEntry object that refers to the member or group by its controller assigned id (which should not be confused with the P4 compiler assigned ID for the action_selector).

A controller api would add group table entries as follows (canned syntax):
add_action_profile_member <action_profile_id>, <controller_assigned_member_id>, <action_id>, <action params ...>
add_action_profile_group <action_profile_id>, <controller_assigned_group_id>, <controller_assigned_member_ids....>
add_table_entry <table_id>, <match_key...>, <controller_assigned_group_id>

If a packet matches on the table_entry added above, it would execute the action_selector algorithm on the group, which would pick the action profile member from the group's members, and execute the corresponding action, thereby increment the direct counter associated with the table_entry.
 
-----------------------------------------------------------------
control foo_ctrl(inout parsed_packet_t hdr) {
  direct_counter(CounterType.packets) foo_counter;
  action foo() {
    foo_counter.count();
  }
  table foo_table {
    key = {
      hdr.ipv4_base.dst_addr : lpm;
      hdr.ipv4_base.src_addr : selector;
    }
    actions = {
      foo;
    }
    counters = foo_counter;
    implementation = action_selector(HashAlgorithm.crc16, 32w1024, 32w14);
  }
  apply {
    foo_table.apply();
  }
}